[JIT] Normalize op aliases

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38746 [JIT] Factor out aliases to separate test
* **#38735 [JIT] Normalize op aliases**
* #38734 [JIT] Rename canonicalize ops

Follow up to my comment https://github.com/pytorch/pytorch/pull/36597/#issuecomment-613674329

This adds a pass to convert op aliases into a normalized form. Having two ops generated in our IR that do the same thing makes the IR harder for downstream consumers of the IR, such as TorchScript passes but also ONNX, glow, etc. 

Another solution would have been to fix our code generation to only emit `aten::abs` from the start. This seems trickier, and doesn't really buy us much if we still have to expose `aten::absolute` in C++, as @glaringlee of the C++ API thinks we should.

Bike shedding: maybe this should be `CanonicalizeOps` instead

Differential Revision: [D21673108](https://our.internmc.facebook.com/intern/diff/D21673108)

Commit:
[JIT] Normalize op aliases

[ghstack-poisoned]

glaringlee commented:
I previously thought this numpy-compatible project is for logical compatible, so instead of adding 'absolute', we can reuse 'abs' and fix the non-numpy-compatible stuff inside if any. Now I see that we also need naming compatible, so adding normalization layer and support alias is very reasonable to me now. 👍 Thanks.

mruberry commented on test/test_jit.py:
```diff
@@ -15903,7 +15917,6 @@ def test_multiple(x: int):
 
         graph = torch.jit.script(test_multiple).graph
         FileCheck().check_count("prim::If", 3, exactly=True).run(graph)
-        print(torch.jit.script(test_multiple).code)
```
Thank you for cleaning this up!

mruberry commented on test/test_jit.py:
```diff
@@ -9092,6 +9092,20 @@ def test_irparser(self):
         """
         FileCheck().run(graph_str, parse_ir(graph_str))
 
+    def test_norm_aliases(self):
```
This test is cool but either in this PR or a follow-up we should make it extensible just like the method of registering aliases in the JIT is. That way when people add a new alias they can also easily extend this test.

eellison replied:
Done here: https://github.com/pytorch/pytorch/pull/38746

mruberry commented:
This is super cool, thanks @eellison! I left a few comments requesting clarification on how to add and test new aliases.

Commit:
Update on "[JIT] Normalize op aliases"


Follow up to my comment https://github.com/pytorch/pytorch/pull/36597/#issuecomment-613674329

This adds a pass to convert op aliases into a normalized form. Having two ops generated in our IR that do the same thing makes the IR harder for downstream consumers of the IR, such as TorchScript passes but also ONNX, glow, etc. 

Another solution would have been to fix our code generation to only emit `aten::abs` from the start. This seems trickier, and doesn't really buy us much if we still have to expose `aten::absolute` in C++, as @glaringlee of the C++ API thinks we should.

Bike shedding: maybe this should be `CanonicalizeOps` instead

[ghstack-poisoned]

mruberry commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -0,0 +1,61 @@
+namespace torch {
+namespace jit {
+
+static const std::unordered_map<Symbol, Symbol> alias_map = {
```
Sorry, my comment here didn't get saved, I guess. What about a comment here explaining how people should add (and test) new aliases?

eellison replied:
I added a comment. I would prefer not to put testing details in source code because it makes our code more verbose and the details would likely grow stale. If someone wants to look at how something is tested the typical thing is to look at git blame or ask around.

mruberry replied:
Sounds good.

Commit:
Update on "[JIT] Normalize op aliases"


Follow up to my comment https://github.com/pytorch/pytorch/pull/36597/#issuecomment-613674329

This adds a pass to convert op aliases into a normalized form. Having two ops generated in our IR that do the same thing makes the IR harder for downstream consumers of the IR, such as TorchScript passes but also ONNX, glow, etc. 

Another solution would have been to fix our code generation to only emit `aten::abs` from the start. This seems trickier, and doesn't really buy us much if we still have to expose `aten::absolute` in C++, as @glaringlee of the C++ API thinks we should.

Bike shedding: maybe this should be `CanonicalizeOps` instead

[ghstack-poisoned]

jamesr66a commented on torch/csrc/jit/passes/normalize_ops.h:
```diff
@@ -0,0 +1,16 @@
+namespace torch {
+namespace jit {
+
+// This passes converts aten ops to a normalized form. It is
```
nit: `This pass` :p

jamesr66a commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -1219,8 +1220,10 @@ void initJitScriptBindings(PyObject* module) {
+
         auto cu = get_python_cu();
         auto name = c10::QualifiedName(qualname);
+        NormalizeOps(graph);
```
Is this also needed in `_create_method_from_trace`? Can we just embed this call in `trace()` (called by `createGraphByTracing`)?  Maybe here:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/tracer.cpp#L454

jamesr66a commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -0,0 +1,65 @@
+    {aten::absolute_, aten::abs_},
+};
+
+void replaceNodeWithNewSymbol(Node* node, Symbol new_symbol) {
```
Would this be a useful method to expose on `Node` or `Graph`? I can definitely think of places I've wanted a function like this elsewhere

eellison replied:
Do you have any examples of other call sites ?

jamesr66a commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -0,0 +1,65 @@
+// having multiple ops in our IR that do the same thing makes the IR more
+// difficult to consumer for downstream user of the IR, such as our own
+// optimization passes here, we convert op aliases into a standard form
+bool normalizeOpAliases(graph_node_list_iterator& iter) {
```
static, or just wrap all the implementation code in an anonymous namespace

jamesr66a reviewed (approved):
Cool!

Do we have a plan for extending this to other use cases? Main one is e.g. `aten::_convolution` v.s. `aten::conv*d`

eellison commented:
@jamesr66a no concrete plan to move over `_convolution` but it is my intention that this pass will eventually encompass that & other use cases

Commit:
Update on "[JIT] Normalize op aliases"


Follow up to my comment https://github.com/pytorch/pytorch/pull/36597/#issuecomment-613674329

This adds a pass to convert op aliases into a normalized form. Having two ops generated in our IR that do the same thing makes the IR harder for downstream consumers of the IR, such as TorchScript passes but also ONNX, glow, etc. 

Another solution would have been to fix our code generation to only emit `aten::abs` from the start. This seems trickier, and doesn't really buy us much if we still have to expose `aten::absolute` in C++, as @glaringlee of the C++ API thinks we should.

Bike shedding: maybe this should be `CanonicalizeOps` instead

[ghstack-poisoned]
